Make new GameState class, minor refactor

ankith26 opened the pull request.

ankith26 committed:
Make new GameState class, minor refactor

Grimmys commented on src/entities/ball.py:
```diff
@@ -1,25 +1,33 @@
+
+    def __init__(self, game_state: GameState):
+        """
+        Initialise Ball class
```
typo: initialize I'm right?

ankith26 replied:
Ohh well I guess it is another american-british english thing. But yea "initialize" is more common, so I guess we should switch to it

Grimmys replied:
Oh no no no, if the word exists in some English local variation it's ok!
Sorry lol

ankith26 replied:
No worries lol

Grimmys commented on src/entities/game_object.py:
```diff
@@ -1,21 +1,37 @@
-    def render(self, screen):
-        if self.image:
+    def render(self, screen: pygame.Surface):
+        if self.image is not None and self.rect is not None:
```
Do we really need `is not None` instead of `if self.image and self.rect:`?
I'm curious to learn new things.

ankith26 replied:
Yea, `if self.rect` just truth-tests the variable. This means that it is `False` if `None`, and if it is a `Rect` object, it uses the `__bool__` method to determine the truth value of the object. Clearly, we want to test for an explicit `None` here, and not the truth value

Grimmys commented on src/entities/laser.py:
```diff
@@ -1,13 +1,21 @@
-        rect = image.get_rect()
-        rect.center = spaceship.rect.center
-        super().__init__(game, image, rect)
+    def __init__(self, game_state: GameState, spaceship: Spaceship):
```
I'm worried by the fact we don't call the parent init method anymore for this object and all others in fact... I fear that we are breaking a key point of OOP lol, and moreover we have duplicate code (for example `self.game_state = game_state` is literally the same statement than in parent init method).
Maybe I can try something to fix that.

ankith26 replied:
Yea right, but I made it this way so that the subclass init does everything the parent init is doing anyways, so the super init call would be redundant. But maybe you are right, for more clarity sake, I should be calling super init

Grimmys replied:
Well, no, it doesn't have to be redundant.
In fact, it is the call to the super init that avoids to have redundancy in each object initialization!
You can check what I pushed, maybe we can even remove the `rect` argument because we don't use it... I found it weird at first glance, I'm always used to simply pass the `image` as an argument and compute the `rect` according to the `image`.

ankith26 replied:
Yea, your new commit fixes things up pretty well, nice!
(and as discussed, we need to make image required, and also use `get_rect` for getting rect internally from image rather than argument)

Grimmys commented on src/entities/paddle.py:
```diff
@@ -1,18 +1,20 @@
+        self.image = game_state.assets.paddle
+        self.rect = self.image.get_rect()
+        self.rect.center = (const.WIDTH // 2, const.HEIGHT - 50)
+        self.game_state = game_state
```
More I compare the init implementation of each object (Paddle, Ball, Laser etc.) more I find them really similar, I'm sure we can do better hehe.

Grimmys commented on src/entities/game_object.py:
```diff
@@ -1,21 +1,37 @@
+    ):
         self.image = image
         self.rect = rect
+        self.game_state = game_state
```
Should game objects really be aware of the game state and store it as an attribute?
It seems weird to me, the game scene should definitely have a reference to the game state and should handling its evolution, but the game objects should not have a control other it, if they need to know the score (for example), the scene can provide them the current value as an argument to a specific method.
It's only my pov.

ankith26 replied:
Yea, I am not particularly happy with this system too. But in cases when a game object needs to spawn a new object, or maybe increase the score, it needs access to the game state

Grimmys replied:
Okay, I still find it weird (for example, if we want to increase the score after the ball hits a brick... In which world would it be ball duty (or brick duty??) to do it, lol), but I understand your point. It's a detail, maybe it's simpler to do it that way. =)

Grimmys committed:
Replace common game object init code by call to the parent init

Grimmys committed:
Changed image argument in game object to be required instead of optional

ankith26 committed:
add more docstrings

Grimmys committed:
Introduced the notion of block size

Added new constants for block size, and brick size that is for now equal to block size.
Switch brick width from 50 to 30.
Added the possibility to rescale an image.

Grimmys committed:
Added spaceship size

Added new constants for spaceship size.
Switch spaceship width from 80 to 60.
Remove space between spaceships.
